wire-ios/Wire-iOS/Sources/Helpers/syncengine/Conversation+Participants.swift

```diff
@@ -38,66 +39,110 @@ extension ZMConversation {
     }
 
     func addOrShowError(participants: [UserType]) {
-        guard let session = ZMUserSession.shared(),
-                session.networkState != .offline else {
-            self.showAlertForAdding(for: NetworkError.offline)
-            return
+        guard
+            let session = ZMUserSession.shared(),
+            session.networkState != .offline
+        else {
+            return showAlertForAdding(for: ConversationError.offline)
         }
 
-        addParticipants(participants) { (result) in
-            switch result {
-            case .failure(let error):
-                self.showAlertForAdding(for: error)
-            default:
-                break
+        let users = participants.materialize(in: session.viewContext)
+        let syncContext = session.syncContext
+        let service = ConversationParticipantsService(context: syncContext)
+
+        Task {
+            do {
+                let users = await syncContext.perform {
+                    users.compactMap {
+                        ZMUser.existingObject(for: $0.objectID, in: syncContext)
+                    }
+                }
+
+                let conversation = try await syncContext.perform { [self] in
+                    return try ZMConversation.existingObject(for: self.objectID, in: syncContext)
+                }
+
+                try await service.addParticipants(users, to: conversation)
+            } catch {
+                await MainActor.run {
+                    self.showAlertForAdding(for: error)
+                }
             }
         }
     }
 
     func removeOrShowError(participant user: UserType, completion: ((VoidResult) -> Void)? = nil) {
-        guard let session = ZMUserSession.shared(),
-            session.networkState != .offline else {
-            self.showAlertForRemoval(for: NetworkError.offline)
-            return
+
+        @Sendable func fail(with error: Error) {
+            showAlertForRemoval(for: error)
+            completion?(.failure(error))
+        }
+
+        guard
+            let session = ZMUserSession.shared(),
+            session.networkState != .offline
+        else {
+            return fail(with: ConversationError.offline)
+        }
+
+        guard let user = user.materialize(in: session.viewContext) else {
```

Comment:
what does this method do?

Reply:
it materialises a list of objects conforming to `UserType` into `ZMUser` either by fetching them from the store if they exist, or by creating them.